torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -373,14 +373,45 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
   threadPool_.run(std::bind(
       [&](RecvWork& work) {
         torch::Tensor& payload = work.payload_;
```

Comment:
Can we move this entire code block to a function `handleRecv(RecvWork)` and then do something like this in this function:

```
try {
  handleRecv(work);
} catch (std::exception& e) {
  // Based on whether work is a request/response we handle the exception appropriately.
}

I feel this would be much cleaner than having individual try-catch blocks spread throughout this function.

Reply:
From a glance, one might think so. But practically speaking, there are roughly 3 specific classes of errors that I believe we want to handle here, in different ways:

1.  Deserialization errors - this is what initially motivated the change. In this case, for server-side we want to return an exception response to the client, or on client-side, just to fulfill the future with an exception.

2. server-side processing errors- suppose the request_callback_impl throws an exception. Thrift handles this ok currently, but process group should catch and channel this back into an exception for the client to see (like Thrift basically does).

3. errors fulfilling the future- this one shouldn't really happen, but Rohan pointed as a possibility, and it's probably best to just log when the added callback itself throws so we can address.

Reply:
Thanks for comments, I'll see if I can add a few test cases somewhere.
I'm debating where to put? Is it acceptable to make fb-only tests here (which would be somewhat simpler), or should they be in OSS?

Reply:
I feel a large try-catch block is safer since it ensures we don't miss out on any cases as Rohan pointed out here https://github.com/pytorch/pytorch/pull/31299#discussion_r359042760. Also, as we modify the logic here in the future its safer to have one try-catch block at the top level which handles all errors. Otherwise, it's easy for developers in the future to add some functionality but they forget to handle exceptions. Also, the single try-catch block won't just catch and log, it should send a response back to the client if its a request and mark an appropriate error on the future if it is a response.

I think the tests should be in OSS since this is OSS code we're modifying.

Reply:
Let me think about how to add coverage here - it's fair to ask for it, though OSS cpp code coverage is quite poor vs. fb-only.

In any case, having an overall try/catch block for 3 (and maybe 2), might well be reasonable. For the  1. case above, I think there's a good argument for more specialized handling. In any case, let's see when we can get this up to parity with Thrift.

Reply:
@jjlilley For now, we (unfortunately) don't have cpp tests for process group agent, there's an issue to add them: https://github.com/pytorch/pytorch/issues/28309 but we probably won't get to this for some time. For now, I'm wondering if you can add them in python similar to what I did in https://github.com/pytorch/pytorch/pull/31023, but simulating an exception on the recv side might be trickier

Reply:
Yeah, thanks, I know process_group_agent has zero cpp test coverage.
I've seen the pattern - people are all quietly hoping somebody else adds it :)

fwiw, I did write up some c++ ProcessGroupAgent benchmarking code in fb-only the other month, when I was trying to optimize rpc perf:
  caffe2/torch/fb/distributed/thriftRpcBackend/test/ProcessGroupAgentBench.cpp

I might try to massage that into something useful for test coverage in OSS.